exchanger/exchanger.go

```diff
@@ -0,0 +1,83 @@
+package exchanger
+
+import (
+	"net"
+	"time"
+
+	"github.com/mesosphere/mesos-dns/logging"
+	"github.com/miekg/dns"
+)
+
+// Exchanger is an interface capturing a dns.Client Exchange method.
+type Exchanger interface {
+	Exchange(m *dns.Msg, addr string) (r *dns.Msg, rtt time.Duration, err error)
+}
+
+// Func is a function type that implements the Exchanger interface.
+type Func func(*dns.Msg, string) (*dns.Msg, time.Duration, error)
+
+// Exchange implements the Exchanger interface.
+func (f Func) Exchange(m *dns.Msg, addr string) (*dns.Msg, time.Duration, error) {
+	return f(m, addr)
+}
+
+// A Decorator adds a layer of behaviour to a given Exchanger.
+type Decorator func(Exchanger) Exchanger
+
+// Decorate decorates an Exchanger with the given Decorators.
+func Decorate(ex Exchanger, ds ...Decorator) Exchanger {
+	decorated := ex
+	for _, decorate := range ds {
+		decorated = decorate(decorated)
+	}
+	return decorated
+}
+
+// Instrumentation returns a Decorator which instruments an Exchanger with the given
+// counter.
+func Instrumentation(c logging.Counter) Decorator {
+	return func(ex Exchanger) Exchanger {
+		return Func(func(m *dns.Msg, a string) (*dns.Msg, time.Duration, error) {
+			defer c.Inc()
+			return ex.Exchange(m, a)
+		})
+	}
+}
+
+// A Recurser returns the addr of the next DNS server to recurse a Msg to. Empty
+// returns signal that further recursion isn't possible or needed.
+type Recurser func(*dns.Msg) string
+
+// Recurse is the default Mesos-DNS Recurser which returns an addr only
```

Comment:
can we clarify that the addresses returned here are nameserver host:port addresses?